packages/starlight/utils/plugins.ts

```diff
@@ -0,0 +1,223 @@
+import type { AstroIntegration } from 'astro';
+import { z } from 'astro/zod';
+import { StarlightConfigSchema, type StarlightUserConfig } from '../utils/user-config';
+import { errorMap } from '../utils/error-map';
+
+/**
+ * Runs Starlight plugins in the order that they are configured after validating the user-provided
+ * configuration and returns the final validated user config that may have been updated by the
+ * plugins and a list of any integrations added by the plugins.
+ */
+export async function runPlugins(
+	starlightUserConfig: StarlightUserConfig,
+	pluginsUserConfig: StarlightPluginsUserConfig,
+	context: StarlightPluginContext
+) {
+	// Validate the user-provided configuration.
+	let userConfig = starlightUserConfig;
+	let starlightConfig = StarlightConfigSchema.safeParse(userConfig, { errorMap });
+
+	if (!starlightConfig.success) {
+		throwValidationError(starlightConfig.error, 'Invalid config passed to starlight integration');
+	}
+
+	// Validate the user-provided plugins configuration.
+	const pluginsConfig = starlightPluginsConfigSchema.safeParse(pluginsUserConfig, {
+		errorMap,
+	});
+
+	if (!pluginsConfig.success) {
+		throwValidationError(
+			pluginsConfig.error,
+			'Invalid plugins config passed to starlight integration'
+		);
+	}
+
+	// A list of Astro integrations added by the various plugins.
+	const integrations: AstroIntegration[] = [];
+
+	for (const {
+		name,
+		hooks: { setup },
+	} of pluginsConfig.data) {
+		await setup({
+			config: pluginsUserConfig ? { ...userConfig, plugins: pluginsUserConfig } : userConfig,
+			updateConfig(newConfig) {
+				// Ensure that plugins do not update the `plugins` config key.
+				if ('plugins' in newConfig) {
+					throw new Error(
+						`The '${name}' plugin tried to update the 'plugins' config key which is not supported.`
+					);
+				}
+
+				// If the plugin is updating the user config, re-validate it.
+				const mergedUserConfig = { ...userConfig, ...newConfig };
+				const mergedConfig = StarlightConfigSchema.safeParse(mergedUserConfig, { errorMap });
+
+				if (!mergedConfig.success) {
+					throwValidationError(
+						mergedConfig.error,
+						`Invalid config update provided by the '${name}' plugin`
+					);
+				}
+
+				// If the updated config is valid, keep track of both the user config and parsed config.
+				userConfig = mergedUserConfig;
+				starlightConfig = mergedConfig;
+			},
+			addIntegration(integration) {
+				// Collect any Astro integrations added by the plugin.
+				integrations.push(integration);
+			},
+			astroConfig: context.config,
```

Comment:
I think as a plugin author I’d expect `astroConfig.integrations` here to include the updated values from previous plugins (or at least to have another way to check this, but here probably make most sense?)

Thoughts on doing something like this? On the one hand it’s slightly lying, but on the other hand, in practice this is what is happening or “will” happen once we update the Astro config after plugins run?

```suggestion
			astroConfig: { ...context.config, integrations: [...context.config.integration, ...integrations] },
```

Effectively mirrors what we do with the `updateConfig()` hook where each plugin is passed `userConfig` as mutated by the previous.

Reply:
Good catch, and I think this is a good solution you’ve come up with. I implemented this and added a test for it.